pytorch/torchtune
# Support Early Exit Loss and/or Layer Dropout

#### Context
What is the purpose of this PR? Is it to
- [X] add a new feature

Implements early exit loss and/pr layer dropout to reproduce experiments in various papers like [LayerSkip](https://arxiv.org/abs/2404.16710), [LITE](https://arxiv.org/abs/2310.18581), [LayerDrop](https://arxiv.org/abs/1909.11556), and [Progressive Layer Dropping](https://arxiv.org/abs/2010.13369).

#### Usage
Download Llama2 7B model:
```
tune download meta-llama/Llama-2-7b-hf --output-dir /tmp/Llama-2-7b-hf --hf-token <HF_TOKEN>
```

To reproduce experiments of various papers that use early exit loss and/or layer dropout:
- [LayerSkip](https://arxiv.org/abs/2404.16710) on TOPv2:
```
tune run --nnodes 1 --nproc_per_node 4 dev/early_exit_finetune_distributed --config recipes/dev/7B_full_early_exit.yaml early_exit_loss.scale=1.0 early_exit_loss.curriculum=torchtune.modules.early_exit_loss.GradualEarlyExitCurriculum early_exit_loss.scale_fn=torchtune.modules.early_exit_loss.linear_l_loss_scale layer_dropout.prob=0.2 layer_dropout.scale=exp
```
- [LITE](https://arxiv.org/abs/2310.18581):
```
tune run --nnodes 1 --nproc_per_node 4 dev/early_exit_finetune_distributed --config recipes/dev/7B_full_early_exit.yaml layer_dropout=null early_exit_loss.layers=8,12,16,20,24,28 early_exit_loss.scale_fn=torchtune.modules.early_exit_loss.uniform_loss_scale early_exit_loss.curriculum=null epochs=5
```

- [LayerDrop](https://arxiv.org/abs/1909.11556):
```
tune run --nnodes 1 --nproc_per_node 4 dev/early_exit_finetune_distributed --config recipes/dev/7B_full_early_exit.yaml early_exit_loss=null layer_dropout.prob=0.2 layer_dropout.layers=1::2
```

- [Progressive Layer Dropping](https://arxiv.org/abs/2010.13369) (The paper also implements a curriculum for layer drop probability which is not yet implemented.):
```
tune run --nnodes 1 --nproc_per_node 4 dev/early_exit_finetune_distributed --config recipes/dev/7B_full_early_exit.yaml early_exit_loss=null layer_dropout.prob=0.5 layer_dropout.scale=exp
```

#### Changelog
- Modified `TransformerDecoder` by refactoring the operations that happen after running all transformer layers that transform last hidden state to a probability vector across all tokens into a separate `unembed()` function. This was needed to re-use `unembed()` to obtain predictions at intermediate layers, and hence calculate losses at intermediate layers.
- Added to `torchtune/modules/common_utils.py` a `slice_str_to_array()` function that can convert numpy-like indexing string into a boolean array.
- Added `torchtune/modules/early_exit_loss.py` to implement early exit loss
- Added `torchtune/modules/layer_dropiut.py` to implement layer dropout
- Added separate recipe script and configuration file
- Added unit test cases


#### Test plan
1. Download Llama 2 7B:
```
tune download meta-llama/Llama-2-7b-hf --output-dir /tmp/Llama-2-7b-hf 
```

2. Finetune on TOPv2 instruction set (a small dataset that we can quickly finetune with early exit loss). W
```
tune run --nnodes 1 --nproc_per_node 8 dev/early_exit_finetune_distributed --config recipes/dev/7B_full_early_exit.yaml 
```
We will run different variants of this command in the Table below to show different speed ups.

3. Convert trained checkpoint to HF naming:
```
mv /tmp/Llama-2-7b-hf/hf_model_0001_0.pt /tmp/Llama-2-7b-hf/pytorch_model-00001-of-00002.bin
mv /tmp/Llama-2-7b-hf/hf_model_0002_0.pt /tmp/Llama-2-7b-hf/pytorch_model-00002-of-00002.bin
```

4. Install HF `transformers` nightly (to enable early exit speculative decoding as described [here](https://huggingface.co/blog/layerskip)):
```
pip install git+https://github.com/huggingface/transformers.git
pip install accelerate
```

5. Save this [code snippet](https://gist.github.com/mostafaelhoushi/bda38661da1738e3a01a07ccf69bde7d) to a file named `eval_hf.py`

6. Run `eval_hf.py` that benchmarks autoregressive decoding and early exit self-speculative decoding:
```
python eval_hf.py
```

You should get an output that looks like this:
```
[INST]Set alarm for 6am every day[\INST]Set alarm for 6 am every day [IN:CREATE_ALARMSet alarm for 6 am every day ]
Orig Time: 1.159
[INST]Set alarm for 6am every day[\INST]Set alarm for 6 am every day [IN:CREATE_ALARMSet alarm for 6 am every day ]
Layerskip Time: 0.500
```
Hence, LayerSkip's training recipe has enabled LayerSkip's (early exit self-speculative decoding) inference that leads to a speedup of 1.159/0.500 =  2.318x.

We ran different variants of training and obtained the results below. Please note that differences in generation time for different training configurations may be due to different number of tokens generated. Hence, to measure speedup, we need to look at the different generation times for the same training configuration:
[Link to Results Spreadsheets](https://docs.google.com/spreadsheets/d/1RrJFKGEeBgndoDKhH6DBl5SouFT-vIaMmvWaYbDc5yk/edit?usp=sharing)

#### Checklist

- [x] run pre-commit hooks and linters (make sure you've first installed via `pre-commit install`)
- [x] add unit tests for any new functionality
- [x] update docstrings for any new or updated methods or classes
- [x] run unit tests via `pytest tests`
- [x] run recipe tests via `pytest tests -m integration_test`
- [x] manually run any new or modified recipes with sufficient proof of correctness
	- [x] include relevant commands and any other artifacts in this summary (pastes of loss curves, eval results, etc.)


#### Future Work

In the future we can:
- **Implement early exit self-speculative decoding in `torchtune/generation/_generation.py`**: `_generation.py` uses similar code to `gpt-fast`. We have already implemented early-exit self-speculative decoding on gpt-fast [here](https://github.com/pytorch-labs/gpt-fast/compare/main...LayerSkip) so it should be straightforward to port to torchtune.
- **Enable users to add their own heads or adapter modules at early exits** rather than using a shared head for all early exits (this can enable implementing a paper like [Kangaroo](https://arxiv.org/abs/2404.18911)).
- **Make `early_exit_loss()` function support generic loss preprocessing**: different training scripts may perform different pre-processing of logits and labels. We need to refactor training scripts slightly to encapsulate such preprocessing into a function to invoke when calculating early exit loss.
- **Make `early_exit_loss()` function support generic loss functions**: support other types of losses like KL divergence loss, or losses represented as lists (as is the case with model's `num_output_chunks`>0).
